### PR TITLE
[Qwen3-Omni Perf]: expose encoder batch wait configuration

### PR DESCRIPTION
**Pull request by cuzmi**
## Motivation

Qwen3-Omni image and audio encoders previously relied on a hard-coded 50 ms batch-wait window. This PR makes the wait configurable through `sgl-omni serve` and changes the Qwen3-Omni default to 10 ms based on image- and audio-input benchmarks.

## Modifications

- Add `--max-batch-wait-ms` to `sgl-omni serve` and apply it to both Qwen3-Omni image and audio encoders.
- Set the default batch-wait window for both Qwen3-Omni encoders to 10 ms.
- Share a streaming chat-response parser between the MMMU and MMSU benchmarks to support TTFT/TTFA measurement for both image and audio inputs.
- Add unit tests covering the 10 ms default, CLI propagation to both encoders, runtime overrides, and negative-value validation.

## Related Issues

Related to #1147.

## Accuracy Test

N/A — no model-side changes.

### Image-input benchmark (MMMU)

Setup:

- Benchmark: `benchmarks/eval/benchmark_omni_mmmu.py`
- Hardware: NVIDIA H100 80GB
- Dataset: `zhaochenyang20/mmmu-ci-50`
- Input modality: image
- Output modalities: streaming text+audio (`--stream --enable-audio`)
- Requests: 50 per run
- Initial sweep: `{0, 1, 5, 10, 50}` ms at c1/c8/c16
- Decision repeats: `{1, 10, 50}` ms, three runs at c1/c16

The table reports the median of the three repeated runs:

| C | Wait | TTFT p50 | TTFT p95 | TTFA p50 | QPS |
|---:|---:|---:|---:|---:|---:|
| 1 | 1 ms | **0.188 s** | 1.325 s | 3.039 s | **0.039** |
| 1 | 10 ms | 0.197 s | **1.263 s** | **3.012 s** | **0.039** |
| 1 | 50 ms | 0.249 s | 1.285 s | 3.054 s | **0.039** |
| 16 | 1 ms | 0.897 s | 1.807 s | 8.737 s | 0.145 |
| 16 | 10 ms | **0.868 s** | **1.609 s** | **8.172 s** | **0.155** |
| 16 | 50 ms | 0.982 s | 6.719 s | 9.970 s | 0.144 |

Compared with the previous 50 ms default, 10 ms provides:

- c1: 20.8% lower TTFT p50 with unchanged QPS;
- c16: 11.6% lower TTFT p50, 18.0% lower TTFA p50, and 7.6% higher QPS.

In the single-run initial sweep, 5 ms showed strong text latency, achieving the best c8 TTFT p50 (0.420 s) and c16 TTFT p95 (2.040 s). However, its c16 TTFA p50 was worse than 10 ms (10.977 s versus 8.967 s), and its c16 QPS was slightly lower (0.150 versus 0.154). Overall, 10 ms offered the best balance between text latency, audio latency, and throughput for this workload.

To verify that the 10 ms versus 50 ms difference is larger than the run-to-run variation, the corresponding run-level results are shown below.

| C | Wait | Run | TTFT p50 | TTFA p50 | QPS |
|---:|---:|---:|---:|---:|---:|
| 1 | 10 ms | 1 | 0.1969 s | 3.0117 s | 0.039 |
| 1 | 10 ms | 2 | 0.1959 s | 3.0696 s | 0.040 |
| 1 | 10 ms | 3 | 0.2097 s | 2.9839 s | 0.038 |
| 1 | 50 ms | 1 | 0.2486 s | 3.0388 s | 0.038 |
| 1 | 50 ms | 2 | 0.2335 s | 3.0961 s | 0.040 |
| 1 | 50 ms | 3 | 0.2606 s | 3.0539 s | 0.039 |
| 16 | 10 ms | 1 | 0.8943 s | 8.3501 s | 0.157 |
| 16 | 10 ms | 2 | 0.8419 s | 7.9610 s | 0.155 |
| 16 | 10 ms | 3 | 0.8681 s | 8.1722 s | 0.151 |
| 16 | 50 ms | 1 | 0.9213 s | 9.1273 s | 0.144 |
| 16 | 50 ms | 2 | 1.0433 s | 10.5204 s | 0.141 |
| 16 | 50 ms | 3 | 0.9820 s | 9.9702 s | 0.146 |

Across all three c16 repeats, 10 ms consistently outperformed 50 ms in TTFT p50, TTFA p50, and QPS. At c1, 10 ms also consistently reduced TTFT p50, while TTFA and QPS remained effectively neutral.

### Audio-input benchmark (MMSU)

To validate the new default with actual audio inputs, I also ran the MMSU benchmark using streaming text+audio output.

Setup:

- Benchmark: `benchmarks/eval/benchmark_omni_mmsu.py`
- Hardware: NVIDIA H100 80GB
- Dataset: `zhaochenyang20/mmsu-ci-2000`
- Input modality: audio
- Output modalities: streaming text+audio (`--modalities text+audio --stream`)
- Requests: 50 per run
- Warm-up requests: 5 per case
- Maximum output tokens: 32
- Batch-wait values: `{10, 50}` ms
- Concurrency levels: `{1, 16}`
- Repeats: four alternating-order runs

All 800 requests completed successfully and returned audio.

| C | Wait | TTFT | TTFT P50 | TTFA | TTFA P50 | QPS |
|---:|---:|---:|---:|---:|---:|---:|
| 1 | 10 ms | **0.172 s** | **0.183 s** | **0.365 s** | **0.367 s** | **2.101** |
| 1 | 50 ms | 0.212 s | 0.221 s | 0.404 s | 0.406 s | 1.912 |
| 16 | 10 ms | **0.387 s** | **0.368 s** | **1.524 s** | 1.541 s | **6.360** |
| 16 | 50 ms | 1.374 s | 0.417 s | 2.530 s | **1.512 s** | 5.583 |

At concurrency 1, the 10 ms wait improves TTFT, TTFA, and throughput. At concurrency 16, TTFA p50 remains comparable, while the 10 ms wait improves TTFT and QPS.

Together, the MMMU image-input and MMSU audio-input results support 10 ms as a better default for both Qwen3-Omni media encoders.

## Checklist

- [x] Format code with pre-commit.
- [x] Add unit tests.
- [x] Update CLI help/docstrings.
- [x] Provide image- and audio-input benchmark results.

### CODE REVIEW

**Review by edwingao28 — CHANGES_REQUESTED** [review]
Thanks for the thorough sweep 

One thing before merge: the issue's step 3 was to propose the optimal value as the new default, not just expose the knob. Your own numbers show 10 ms beating 50 ms everywhere: 16 TTFT p50 −11.6%, TTFA −18%, QPS +7.6%, and no regression at c1. With data this one-sided, keeping 50 as the default locks the win behind a flag most users will never touch.

1. change the default to 10ms (keep the override)
2. update the doc line to reflect the new default
3. paste the rawnumber of three repeated runs, just want to confirm the run to run spread is simaller than the 10 verse 50 gap.

**Comment by cuzmi** [comment]
Hi @edwingao28, I've pushed the requested changes. The default is now set to 10 ms while keeping the override available, and the documentation has been updated accordingly.

The raw run-level results for the 10 ms and 50 ms decision points are shown below. Each run completed 50/50 requests without failures.

| C | Wait | Run | TTFT p50 | TTFA p50 | QPS |
|---:|---:|---:|---:|---:|---:|
| 1 | 10 ms | 1 | 0.1969 s | 3.0117 s | 0.039 |
| 1 | 10 ms | 2 | 0.1959 s | 3.0696 s | 0.040 |
| 1 | 10 ms | 3 | 0.2097 s | 2.9839 s | 0.038 |
| 1 | 50 ms | 1 | 0.2486 s | 3.0388 s | 0.038 |
| 1 | 50 ms | 2 | 0.2335 s | 3.0961 s | 0.040 |
| 1 | 50 ms | 3 | 0.2606 s | 3.0539 s | 0.039 |
| 16 | 10 ms | 1 | 0.8943 s | 8.3501 s | 0.157 |
| 16 | 10 ms | 2 | 0.8419 s | 7.9610 s | 0.155 |
| 16 | 10 ms | 3 | 0.8681 s | 8.1722 s | 0.151 |
| 16 | 50 ms | 1 | 0.9213 s | 9.1273 s | 0.144 |
| 16 | 50 ms | 2 | 1.0433 s | 10.5204 s | 0.141 |
| 16 | 50 ms | 3 | 0.9820 s | 9.9702 s | 0.146 |

**Comment by edwingao28** [comment]
@cuzmi please update your raw level results and new deault value to pr body as well

**Comment by cuzmi** [comment]
> @cuzmi please update your raw level results and new deault value to pr body as well

Done — updated the raw results and the new default in the PR body.

File: sglang_omni/cli/serve.py [hunk 1]
```diff
@@ -377,6 +377,33 @@ def apply_encoder_mem_reserve_cli_override(
...
+        for stage in pipeline_config.stages
+        if stage.name in {"image_encoder", "audio_encoder"}
+        and "max_batch_wait_ms" in (stage.factory_args or {})
+    ]
+    if not encoder_stages:
+        _raise_unsupported_flag(pipeline_config, "--max-batch-wait-ms")
+
+    for stage in encoder_stages:
+        factory_args = dict(stage.factory_args or {})
+        factory_args["max_batch_wait_ms"] = int(max_batch_wait_ms)
```
Comment (SandyLuXY) [inline sglang_omni/cli/serve.py]: This updates only `factory_args`, but `resolve_stage_static_factory_args()` merges `runtime_overrides` afterward, so a config with encoder `max_batch_wait_ms: 50` there still launches at 50 after `--max-batch-wait-ms 5`. Please update both sources.

File: sglang_omni/models/qwen3_omni/config.py [hunk 2]
```diff
@@ -83,7 +87,11 @@ def _audio_encoder_stage(*, gpu: int, process: str) -> StageConfig:
         name="audio_encoder",
         process=process,
         factory=f"{_PKG}.stages.create_audio_encoder_executor",
-        factory_args={"device": "cuda", "dtype": None},
+        factory_args={
+            "device": "cuda",
+            "dtype": None,
+            "max_batch_wait_ms": 10,
```
Comment (SandyLuXY) [inline sglang_omni/models/qwen3_omni/config.py:93]: This changes the audio encoder's default wait from 50 ms to 10 ms, but all reported measurements use MMMU, whose request builder sends only the sample's `images`. `--enable-audio` changes the requested output modalities and activates the talker/code2wav path

File: sglang_omni/cli/serve.py (hunk 1 above)
Comment (cuzmi) [reply sglang_omni/cli/serve.py]: Fixed — the CLI override now updates both `factory_args` and `runtime_overrides`. I also added a test for the 50 → 5 ms override case.

File: sglang_omni/models/qwen3_omni/config.py (hunk 2 above)
Comment (cuzmi) [reply sglang_omni/models/qwen3_omni/config.py:93]: Thanks for catching this. I added streaming TTFT/TTFA to MMSU and reran with actual audio inputs.

On an H100 with 50 samples per case, 10 ms vs. 50 ms gave c1 TTFT/TTFA p50 of 0.183/0.367 s vs. 0.221/0.406 s. At c16, TTFT p50 was 0.368 s vs. 0.417 s and QPS was 6.360 vs. 5.583; TTFA p50 was about the same.

All 800 requests succeeded and returned audio. Full results are in the PR body.
